### PR TITLE
ROOT start up time and memory optimization

### PR DESCRIPTION
**Pull request by yamaguchi1024**
There are some codes compiled at the start up time. For example,
    - #include \"cling/Interpreter/RuntimeUniverse.h\"
    - #include \"cling/Interpreter/DynamicLookupRuntimeUniverse.h\"
    - namespace cling { class Interpreter; namespace runtime { Interpreter* gCling }}}
    - PrintValue
  These are passed to Cling as string and initialized at the start up time.  So I think it makes sense to reduce top-level global variables, #includes and eagerly deserializations.

  1. Global variables
    If we break at emitModule, we can get a list of global variables and functions which are actually deserialized. These include functions, variables, STL classes and all the functions derives from them. I tried to change them to for example constexpr, so that it's processed at compile time.
  2. Eagerly Deserialized decls
    Thanks to @Axel 's hint and tip, we could minimize eagerly deserialized decls deserialized in ASTReader::PassInterestingDeclsToConsumer. We already removed most of eagerly deserialized decls (Some are remaining to be removed, some are hard to remove and some don't cost a lot).

  So far, we got 9.2% of cpu time improvement, 15% real time improvement and 8.8% of memory improvement at start up time in release build. I took an average, but the variation of the real-time was very big.

  - root.exe -q -l
    - master
      cpu time = 0.09186914285714286 sec (average of 7 times)
      real time = 0.18371428571428572 sec (average of 7 times)
      res  memory = 142.008 Mbytes

    - HEAD
      cpu time = 0.08337842857142856 sec
      real time = 0.15685714285714286 sec
      res  memory = 129.508 Mbytes

  - hsimple.C
  Improved by 13% of cpu time and 8.5% of memory
    - master
      cpu time = 0.0954708 sec (average)
      res  memory = 142.891 Mbytes

    - HEAD
      cpu time = 0.0833258 sec
      res  memory = 130.73 Mbytes

  With modules
    - Improvement by 17.7% in cputime and 2% in memory on root.exe -q -l
    (For memory, small improvement is because most of the memory is taken by LoadModules)
    - With this patch, release modules is 11.2% slower in cpu time and 6% better in residential memory compared to release without modules.

### CODE REVIEW

**Comment by dpiparo** [comment]
Hi @yamaguchi1024 , you mention cpu time. Would it be better to focus on real time since we are optimising for a warm body waiting for ROOT to start?

**Comment by dpiparo** [comment]
btw @yamaguchi1024 , amzing! with these minimal changes we save 10MB at startup, cool!

**Comment by yamaguchi1024** [comment]
@dpiparo Thanks! Agree I'll post a real-time comparison as well.

File: interpreter/cling/include/cling/Interpreter/Value.h [hunk 1]
```diff
@@ -10,7 +10,7 @@
 #ifndef CLING_VALUE_H
 #define CLING_VALUE_H
 
-#include <stddef.h>
+#include <cstddef>
```
Comment (Axel-Naumann) [inline interpreter/cling/include/cling/Interpreter/Value.h]: This should go into a separate PR.

File: net/net/inc/TSecContext.h [hunk 2]
```diff
@@ -27,7 +27,7 @@
 #include "TDatime.h"
 
 // Jan 1, 1995, 00:00:00 in sec from EPOCH (Jan 1, 1970)
-const TDatime kROOTTZERO = 788914800;
+constexpr Int_t kROOTTZERO = 788914800;
```
Comment (Axel-Naumann) [inline net/net/inc/TSecContext.h]: You shouldn't change the type of this variable; this changes its meaning and this variable is public / part of the interface.

Instead, please mark it as `extern const` and provide the definition in the source file.

File: tmva/tmva/inc/TMVA/IPruneTool.h [hunk 3]
```diff
@@ -73,8 +73,8 @@ Any tool which implements the interface should provide two modes for tree prunin
 
       typedef std::vector<const Event*> EventSample;
 
-      IPruneTool( );
-      virtual ~IPruneTool();
+      IPruneTool() : fPruneStrength(0.0), S(0), B(0){};
+      virtual ~IPruneTool() {};
```
Comment (Axel-Naumann) [inline tmva/tmva/inc/TMVA/IPruneTool.h]: The destructor is the key function and should be outlined.

File: interpreter/cling/include/cling/Interpreter/DynamicLookupRuntimeUniverse.h [hunk 4]
```diff
@@ -13,7 +13,7 @@
 #error "This file must not be included by compiled programs."
 #endif
 
-#include "cling/Interpreter/DynamicExprInfo.h"
+class DynamicExprInfo;
```
[the remark this answers is not in the document]
Comment (yamaguchi1024) [reply interpreter/cling/include/cling/Interpreter/DynamicLookupRuntimeUniverse.h]: Thanks for comments!I'll make a different PR for headers,

**Comment by yamaguchi1024** [comment]
@phsft-bot build!

**Comment by vgvassilev** [comment]
FWIW, we need to fix this odd behavior in the ASTReader. This PR is a good for ROOT because it eliminates some bad coding practices. However, there is a lot of code out there which we cannot control and makes sense fixing.

The underlying issue is that the clang ASTReader eagerly deserializes vtables and inlines upon PCH/PCM loading time. There is no good reason for doing that and we can switch this code to lazily deserialize the entities on demand.

**Comment by vgvassilev** [comment]
@yamaguchi1024, I only noticed once I merged. Please make sure you have good-looking git commit history next time!
